File: src/main/java/ch/zxseitz/j3de/graphics/mesh/Rectangle2dFactory.java
package ch.zxseitz.j3de.graphics.mesh;

import ch.zxseitz.j3de.graphics.core.PrimitiveType;

public class Rectangle2dFactory {
    private final PrimitiveType mode;
    private final int[] indices;
    private final float[] uv;

    public Rectangle2dFactory() {
        this.mode = PrimitiveType.TRIANGLE_FAN;
        this.indices = new int[] {
                0, 1, 2, 3
        };
        this.uv = new float[] {
                0f, 0f,
                0f, 1f,
                1f, 1f,
                1f, 0f,
        };
    }

    public PrimitiveType getMode() {
        return mode;
    }

    public int[] getIndices() {
        return indices;
    }

    public float[] getPos3(float width, float height) {
        var x = width / 2;
        var y = height / 2;
        return new float[] {
                -x, -y, 0f,
                x, -y, 0f,
                x, y, 0f,
                -x, y, 0f
        };
    }

    public float[] getUv2() {
        return uv;
    }
}

File: src/main/java/ch/zxseitz/j3de/graphics/mesh/Rectangle2dFactory.kt
package ch.zxseitz.j3de.graphics.mesh;

import ch.zxseitz.j3de.graphics.core.PrimitiveType;

class Rectangle2dFactory {
    val mode: PrimitiveType = PrimitiveType.TRIANGLE_FAN
    val indices: IntArray = intArrayOf(0, 1, 2, 3)
    val uv: FloatArray = floatArrayOf(
        0f, 0f,
        0f, 1f,
        1f, 1f,
        1f, 0f,
    )

    fun getPos3(width: Float, height: Float): FloatArray {
        val x = width / 2;
        val y = height / 2;
        return floatArrayOf(
            -x, -y, 0f,
            x, -y, 0f,
            x, y, 0f,
            -x, y, 0f
        );
    }
}
